agent/api/task/task.go

```diff
@@ -3502,3 +3504,138 @@ func (task *Task) IsServiceConnectConnectionDraining() bool {
 	defer task.lock.RUnlock()
 	return task.ServiceConnectConnectionDrainingUnsafe
 }
+
+// ToHostResources will convert a task to a map of resources which ECS takes into account when scheduling tasks on instances
+// * CPU
+//   - If task level CPU is set, use that
+//   - Else add up container CPUs
+//
+// * Memory
+//   - If task level memory is set, use that
+//   - Else add up container level
+//   - If memoryReservation field is set, use that
+//   - Else use memory field
+//
+// * Ports (TCP/UDP)
+//   - Only account for hostPort
+//   - Don't need to account for awsvpc mode, each task gets its own namespace
+//
+// * GPU
+//   - Return num of gpus requested (len of GPUIDs field)
+//
+// TODO remove ToHostResources is used
+//
+//lint:file-ignore U1000 Ignore all unused code
+func (task *Task) ToHostResources() map[string]*ecs.Resource {
+	resources := make(map[string]*ecs.Resource)
+	// CPU
+	if task.CPU > 0 {
+		// cpu unit is vcpu at task level
+		// convert to cpushares
+		taskCPUint64 := int64(task.CPU * 1024)
+		resources["CPU"] = &ecs.Resource{
+			Name:         utils.Strptr("CPU"),
+			Type:         utils.Strptr("INTEGER"),
+			IntegerValue: &taskCPUint64,
+		}
+	} else {
+		// cpu unit is cpushares at container level
+		containerCPUint64 := int64(0)
+		for _, container := range task.Containers {
+			containerCPUint64 += int64(container.CPU)
+		}
+		resources["CPU"] = &ecs.Resource{
+			Name:         utils.Strptr("CPU"),
+			Type:         utils.Strptr("INTEGER"),
+			IntegerValue: &containerCPUint64,
+		}
+	}
+
+	// Memory
+	if task.Memory > 0 {
+		// memory unit is MiB at task level
+		taskMEMint64 := task.Memory
+		resources["MEMORY"] = &ecs.Resource{
+			Name:         utils.Strptr("MEMORY"),
+			Type:         utils.Strptr("INTEGER"),
+			IntegerValue: &taskMEMint64,
+		}
+	} else {
+		containerMEMint64 := int64(0)
+		// To parse memory reservation / soft limit
+		hostConfig := &dockercontainer.HostConfig{}
+
+		for _, c := range task.Containers {
+			if c.DockerConfig.HostConfig != nil {
+				err := json.Unmarshal([]byte(*c.DockerConfig.HostConfig), hostConfig)
+				if err != nil || hostConfig.MemoryReservation <= 0 {
```

Comment:
Just out of curious, is there a case that MemoryReservation actually less than 0? How did we know it's unit with unmarshal?

Reply:
I don't see a scenario where it will be less than 0. Units are mentioned in comments [here](https://github.com/aws/amazon-ecs-agent/blob/e28ada643fa37572031d7f9cf2f208bb52348f93/agent/vendor/github.com/docker/docker/api/types/container/host_config.go#L366), also verified by manual runs.

Reply:
`int64` implies that this is a signed integer (ie could be negative).  You might also add another check to validate that the HostConfig is an int.

Reply:
[MemoryReservation](https://github.com/aws/amazon-ecs-agent/blob/e28ada643fa37572031d7f9cf2f208bb52348f93/agent/vendor/github.com/docker/docker/api/types/container/host_config.go#L366) is the soft limit we are concerned with here, which is part of the HostConfig. I believe we are already checking if it is zero or negative by the check here `hostConfig.MemoryReservation <= 0`, so should be good imo.

Reply:
And that MemoryReservation field in the HostConfig is zero-initialized by default so we'll always have it if we don't have a null HostConfig.  Resolving.

Reply:
Yes that's right